x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IngestPipelineConfig.java

```diff
@@ -17,11 +17,11 @@
 /**
  * Describes an ingest pipeline to be loaded from a resource file for use with an {@link IndexTemplateRegistry}.
  */
-public class IngestPipelineConfig {
-    private final String id;
-    private final String resource;
-    private final int version;
-    private final String versionProperty;
+public abstract class IngestPipelineConfig {
```

Comment:
I never thought of subclassing as heavyweight, but as a very basic language feature.

> The only differences between the other subclasses is which XContent type it returns (could be a member variable passed in via the constructor instead) and how the data is loaded (could be specified with a supplier function when the config is created instead of needing a subclass).

To me this sounds like you prefer to implement polymorphism while avoiding the language's most basic concept for polymorphism (inheritance). I can't see how what you propose is better that inheritance, especially in terms of readability, with class names indicate exactly what they differ at and method overloads. However, if there is tendency in this code base to avoid class hierarchies when possible, I will follow suit.

Reply:
Maybe if I rename the `ApmIngestPipelineConfig` to `YamlIngestPipelineConfig` it would make more sense. I prefixed it with "Apm" as it is located within the APM plugin and currently doesn't serve as a general-purpose subclass, but naming it consistently with the other subclass (`JsonIngestPipelineConfig`) will probably clarify more what this class hierarchy is about.

Reply:
In my experience it is better to value composition over inheritance, especially in cases such as these where the subclasses are primarily providing values to the main class instead of providing substantial changes in behavior.

For instance, the subclasses each have a method that returns a constant value of which XContent they are meant to handle. That constant could just be passed in as part of the object's constructor. The most significant difference is how they load data, and even that is just a small difference in which format the data is provided in. That could easily be abstracted out of the class with a function or supplier instead of creating an object hierarchy on the ingest pipeline config.

Ultimately it's probably not worth a long discussion. My feedback is admittedly a nitpick, and my concern is primarily structural instead of how the code is behaving. I tend to have a really suspicious response to changes that introduce an object hierarchy.

Reply:
I see what you're saying and I generally agree.
I [reverted](https://github.com/elastic/elasticsearch/pull/97546#discussion_r1297064084) @axw's change of this class to make sure it stays with a single functionality, which is knowing how to load the underlying file. Maybe this is why we see it differently - to me this is ALL these classes are doing. 
Everything in core uses JSON pipeline files that are served from the same location (which I think is a limitation by itself), so I think it's easier for all users to pick a type instead of specifying the same argument in constructor and a supplier.